buckets() panics in flux repl on algo-w branch

Seems like we need a way to send calls to `buckets()` to the remote server, or implement a local version of it.
```
$ go run ./cmd/influx repl -o cwolff -t ...
> buckets()
panic: interface conversion: interface {} is nil, not influxdb.StorageDependencies

goroutine 1 [running]:
github.com/influxdata/influxdb/query/stdlib/influxdata/influxdb.GetStorageDependencies(...)
	/Users/cwolff/workspace/github.com/influxdata/influxdb/query/stdlib/influxdata/influxdb/dependencies.go:29
github.com/influxdata/influxdb/query/stdlib/influxdata/influxdb.createBucketsSource(0x631f480, 0x7458918, 0x945cd771fb4a2d34, 0xaa84199846c7979f, 0x63427c0, 0xc0004a24c0, 0xaa84199846c7979f, 0x4438c2a, 0xc00048db80, 0xc0005e6bc0)
	/Users/cwolff/workspace/github.com/influxdata/influxdb/query/stdlib/influxdata/influxdb/buckets.go:112 +0x4ce
github.com/influxdata/flux/execute.(*createExecutionNodeVisitor).Visit(0xc0005e6f60, 0x6355e80, 0xc000353050, 0x74211e0, 0xc000a33460)
	/Users/cwolff/workspace/github.com/influxdata/flux/execute/executor.go:192 +0x533
github.com/influxdata/flux/plan.walk(0x6355e80, 0xc000353050, 0x6147808, 0xc000a33578, 0xc000a333d0, 0x1, 0x2)
	/Users/cwolff/workspace/github.com/influxdata/flux/plan/walk.go:88 +0x19e
github.com/influxdata/flux/plan.walk(0x6355e80, 0xc000352f30, 0x6147808, 0xc000a33578, 0xc000a333d0, 0xc000956780, 0x60)
	/Users/cwolff/workspace/github.com/influxdata/flux/plan/walk.go:82 +0x15d
github.com/influxdata/flux/plan.(*Spec).BottomUpWalk(0xc0004a21c0, 0xc000a33578, 0xc000956780, 0x5282f80)
	/Users/cwolff/workspace/github.com/influxdata/flux/plan/walk.go:60 +0x2fd
github.com/influxdata/flux/execute.(*executor).createExecutionState(0xc0002a4860, 0x6339780, 0xc0008e52c0, 0xc0004a21c0, 0xc00048dbd0, 0x5282f80, 0xc0002a4860, 0x6324180)
	/Users/cwolff/workspace/github.com/influxdata/flux/execute/executor.go:104 +0x18c
github.com/influxdata/flux/execute.(*executor).Execute(0xc0002a4860, 0x6339780, 0xc0008e52c0, 0xc0004a21c0, 0xc00048dbd0, 0x7, 0x0, 0x0, 0x0)
	/Users/cwolff/workspace/github.com/influxdata/flux/execute/executor.go:70 +0x5d
github.com/influxdata/flux/lang.(*Program).Start(0xc0008e5230, 0x6339780, 0xc0008e5260, 0xc00048dbd0, 0xc0008e5260, 0x6307080, 0xc0008e5230, 0x0)
	/Users/cwolff/workspace/github.com/influxdata/flux/lang/compiler.go:253 +0x1f8
github.com/influxdata/flux/repl.(*REPL).doQuery(0xc00081d680, 0x63396c0, 0xc0004a2140, 0xc000352bd0, 0x6342740, 0xc000051e80, 0x0, 0x0)
	/Users/cwolff/workspace/github.com/influxdata/flux/repl/repl.go:229 +0x205
github.com/influxdata/flux/repl.(*REPL).executeLine(0xc00081d680, 0xc000742e90, 0x9, 0x0, 0xc000a339d8)
	/Users/cwolff/workspace/github.com/influxdata/flux/repl/repl.go:188 +0x372
...
```

--- a/go.mod
+++ b/go.mod
@@ -42,7 +42,7 @@ require (
 	github.com/hashicorp/raft v1.0.0 // indirect
 	github.com/hashicorp/vault/api v1.0.2
 	github.com/influxdata/cron v0.0.0-20191203200038-ded12750aac6
-	github.com/influxdata/flux v0.64.1-0.20200312214513-8048b2f76be0
+	github.com/influxdata/flux v0.64.1-0.20200316205527-19a6b22ea80a
 	github.com/influxdata/httprouter v1.3.1-0.20191122104820-ee83e2772f69
 	github.com/influxdata/influxql v0.0.0-20180925231337-1cbfca8e56b6
 	github.com/influxdata/usage-client v0.0.0-20160829180054-6d3895376368

--- a/query/promql/internal/promqltests/go.mod
+++ b/query/promql/internal/promqltests/go.mod
@@ -7,7 +7,7 @@ require (
 	github.com/glycerine/goconvey v0.0.0-20190410193231-58a59202ab31 // indirect
 	github.com/gogo/protobuf v1.3.0 // indirect
 	github.com/google/go-cmp v0.3.1
-	github.com/influxdata/flux v0.64.1-0.20200312214513-8048b2f76be0
+	github.com/influxdata/flux v0.64.1-0.20200316205527-19a6b22ea80a
 	github.com/influxdata/influxdb v0.0.0-20190925213338-8af36d5aaedd
 	github.com/influxdata/influxql v1.0.1 // indirect
 	github.com/influxdata/promql/v2 v2.12.0

--- a/query/stdlib/influxdata/influxdb/buckets.go
+++ b/query/stdlib/influxdata/influxdb/buckets.go
@@ -15,8 +15,23 @@ import (
 	"github.com/influxdata/influxdb/query"
 )
 
+const BucketsKind = "influxdata/influxdb.localBuckets"
+
 func init() {
-	execute.RegisterSource(influxdb.BucketsKind, createBucketsSource)
+	execute.RegisterSource(BucketsKind, createBucketsSource)
+	plan.RegisterPhysicalRules(LocalBucketsRule{})
+}
+
+type LocalBucketsProcedureSpec struct {
+	plan.DefaultCost
+}
+
+func (s *LocalBucketsProcedureSpec) Kind() plan.ProcedureKind {
+	return BucketsKind
+}
+
+func (s *LocalBucketsProcedureSpec) Copy() plan.ProcedureSpec {
+	return new(LocalBucketsProcedureSpec)
 }
 
 type BucketsDecoder struct {
@@ -99,7 +114,7 @@ func (bd *BucketsDecoder) Close() error {
 }
 
 func createBucketsSource(prSpec plan.ProcedureSpec, dsid execute.DatasetID, a execute.Administration) (execute.Source, error) {
-	_, ok := prSpec.(*influxdb.BucketsProcedureSpec)
+	_, ok := prSpec.(*LocalBucketsProcedureSpec)
 	if !ok {
 		return nil, &flux.Error{
 			Code: codes.Internal,
@@ -128,3 +143,27 @@ type AllBucketLookup interface {
 	FindAllBuckets(ctx context.Context, orgID platform.ID) ([]*platform.Bucket, int)
 }
 type BucketDependencies AllBucketLookup
+
+type LocalBucketsRule struct{}
+
+func (rule LocalBucketsRule) Name() string {
+	return "influxdata/influxdb.LocalBucketsRule"
+}
+
+func (rule LocalBucketsRule) Pattern() plan.Pattern {
+	return plan.Pat(influxdb.BucketsKind)
+}
+
+func (rule LocalBucketsRule) Rewrite(node plan.Node) (plan.Node, bool, error) {
+	fromSpec := node.ProcedureSpec().(*influxdb.BucketsProcedureSpec)
+	if fromSpec.Host != nil {
+		return node, false, nil
+	} else if fromSpec.Org != nil {
+		return node, false, &flux.Error{
+			Code: codes.Unimplemented,
+			Msg:  "buckets cannot list from a separate organization; please specify a host or remove the organization",
+		}
+	}
+
+	return plan.CreateLogicalNode("localBuckets", &LocalBucketsProcedureSpec{}), true, nil
+}

--- a/query/stdlib/influxdata/influxdb/v1/databases.go
+++ b/query/stdlib/influxdata/influxdb/v1/databases.go
@@ -6,66 +6,35 @@ import (
 	"time"
 
 	"github.com/influxdata/flux"
+	"github.com/influxdata/flux/codes"
 	"github.com/influxdata/flux/execute"
 	"github.com/influxdata/flux/memory"
 	"github.com/influxdata/flux/plan"
-	"github.com/influxdata/flux/runtime"
-	"github.com/influxdata/flux/stdlib/influxdata/influxdb/v1"
+	v1 "github.com/influxdata/flux/stdlib/influxdata/influxdb/v1"
 	"github.com/influxdata/flux/values"
 	platform "github.com/influxdata/influxdb"
 	"github.com/influxdata/influxdb/query"
 	"github.com/pkg/errors"
 )
 
-const DatabasesKind = v1.DatabasesKind
-
-type DatabasesOpSpec struct {
-}
-
-func init() {
-	databasesSignature := runtime.MustLookupBuiltinType("influxdata/influxdb/v1", DatabasesKind)
-	runtime.ReplacePackageValue("influxdata/influxdb/v1", DatabasesKind, flux.MustValue(flux.FunctionValue(DatabasesKind, createDatabasesOpSpec, databasesSignature)))
-	flux.RegisterOpSpec(DatabasesKind, newDatabasesOp)
-	plan.RegisterProcedureSpec(DatabasesKind, newDatabasesProcedure, DatabasesKind)
-}
-
-func createDatabasesOpSpec(args flux.Arguments, a *flux.Administration) (flux.OperationSpec, error) {
-	spec := new(DatabasesOpSpec)
-	return spec, nil
-}
-
-func newDatabasesOp() flux.OperationSpec {
-	return new(DatabasesOpSpec)
-}
-
-func (s *DatabasesOpSpec) Kind() flux.OperationKind {
-	return DatabasesKind
-}
-
-type DatabasesProcedureSpec struct {
-	plan.DefaultCost
-}
-
-func newDatabasesProcedure(qs flux.OperationSpec, pa plan.Administration) (plan.ProcedureSpec, error) {
-	_, ok := qs.(*DatabasesOpSpec)
-	if !ok {
-		return nil, fmt.Errorf("invalid spec type %T", qs)
-	}
-
-	return &DatabasesProcedureSpec{}, nil
-}
-
-func (s *DatabasesProcedureSpec) Kind() plan.ProcedureKind {
-	return DatabasesKind
-}
-
-func (s *DatabasesProcedureSpec) Copy() plan.ProcedureSpec {
-	ns := new(DatabasesProcedureSpec)
-	return ns
-}
+const DatabasesKind = "influxdata/influxdb/v1.localDatabases"
 
 func init() {
 	execute.RegisterSource(DatabasesKind, createDatabasesSource)
+	plan.RegisterPhysicalRules(LocalDatabasesRule{})
+}
+
+type LocalDatabasesProcedureSpec struct {
+	plan.DefaultCost
+}
+
+func (s *LocalDatabasesProcedureSpec) Kind() plan.ProcedureKind {
+	return DatabasesKind
+}
+
+func (s *LocalDatabasesProcedureSpec) Copy() plan.ProcedureSpec {
+	ns := new(LocalDatabasesProcedureSpec)
+	return ns
 }
 
 type DatabasesDecoder struct {
@@ -179,7 +148,7 @@ func (bd *DatabasesDecoder) Close() error {
 }
 
 func createDatabasesSource(prSpec plan.ProcedureSpec, dsid execute.DatasetID, a execute.Administration) (execute.Source, error) {
-	_, ok := prSpec.(*DatabasesProcedureSpec)
+	_, ok := prSpec.(*LocalDatabasesProcedureSpec)
 	if !ok {
 		return nil, fmt.Errorf("invalid spec type %T", prSpec)
 	}
@@ -220,4 +189,28 @@ func (d DatabasesDependencies) Validate() error {
 		return errors.New("missing buckets lookup dependency")
 	}
 	return nil
+}
+
+type LocalDatabasesRule struct{}
+
+func (rule LocalDatabasesRule) Name() string {
+	return "influxdata/influxdb.LocalDatabasesRule"
+}
+
+func (rule LocalDatabasesRule) Pattern() plan.Pattern {
+	return plan.Pat(v1.DatabasesKind)
+}
+
+func (rule LocalDatabasesRule) Rewrite(node plan.Node) (plan.Node, bool, error) {
+	fromSpec := node.ProcedureSpec().(*v1.DatabasesProcedureSpec)
+	if fromSpec.Host != nil {
+		return node, false, nil
+	} else if fromSpec.Org != nil {
+		return node, false, &flux.Error{
+			Code: codes.Unimplemented,
+			Msg:  "buckets cannot list from a separate organization; please specify a host or remove the organization",
+		}
+	}
+
+	return plan.CreateLogicalNode("localDatabases", &LocalDatabasesProcedureSpec{}), true, nil
 }

